### PR TITLE
fixing the excessive use of sync on the serial queue (that leads to d…

### PR DESCRIPTION
**Pull request by vodovozovge**
…eadlocks). Also don't notify delegates about the error if the error indicates that the connection is cancelled

### CODE REVIEW

**Review by hoangnami — APPROVED** [review]
Tested.

File: Sources/SwiftCoAP/SCMessage.swift [hunk 1]
```diff
@@ -159,7 +159,9 @@ public final class SCCoAPUDPTransportLayer {
                 return
             }
             if let error = maybeError {
-                self.notifyDelegatesAboutError(for: connection.endpoint, error: error)
+                if error != NWError.posix(.ECANCELED) {
```
Comment (illabo) [inline Sources/SwiftCoAP/SCMessage.swift:162]: This one is neat. Never thought about getting it and not getting canceled state first.

**Review by illabo — APPROVED** [review]
LGTM. 
Sure we don't need the barrier when dealing with arrays?

**Comment by vodovozovge** [comment]
supposedly arrays are thread-safe https://developer.apple.com/swift/blog/?id=10 , so yeah don't need anything special there
>Importantly, you can safely pass copies of values across threads without synchronization.

**Comment by illabo** [comment]
> Importantly, you can safely pass copies of values across threads without synchronization.

I've seen with my meaty eyeballs ~C-beams glitter in the dark near the Tannhauser gate~ the crashes on race write to the array.

**Pull request merged**